File: typescript/src/server/file-queue-protocol-plugin.ts
import { mkdir, readdir, readFile, rename, rm, writeFile } from 'node:fs/promises'
import { basename, join } from 'node:path'
import { HttpError, type RouteContext } from '../types'
import type { PLATServerProtocolPlugin, PLATServerTransportRuntime } from './protocol-plugin'
import type { PLATFileQueueRequest, PLATFileQueueResponse } from './file-queue'
import type { FileQueueOptions } from './config/types'

export interface FileQueueProtocolPluginOptions {
  config: FileQueueOptions | false | undefined
}

export function createFileQueueProtocolPlugin(options: FileQueueProtocolPluginOptions): PLATServerProtocolPlugin {
  if (!options.config || typeof options.config !== 'object') {
    return { name: 'file' }
  }
  const config = options.config

  let runtime: PLATServerTransportRuntime | undefined
  let timer: ReturnType<typeof setInterval> | undefined
  let busy = false

  async function archiveRequest(sourcePath: string): Promise<void> {
    if (config.archive === false) {
      await rm(sourcePath, { force: true })
      return
    }
    if (typeof config.archive === 'string') {
      await rename(sourcePath, join(config.archive, basename(sourcePath)))
      return
    }
    await rm(sourcePath, { force: true })
  }

  async function processOnce(): Promise<void> {
    if (!runtime || busy) return
    busy = true
    try {
      await mkdir(config.inbox, { recursive: true })
      await mkdir(config.outbox, { recursive: true })
      if (config.archive && typeof config.archive === 'string') {
        await mkdir(config.archive, { recursive: true })
      }

      const files = (await readdir(config.inbox))
        .filter((name) => name.endsWith('.json'))
        .sort()

      for (const name of files) {
        const sourcePath = join(config.inbox, name)
        let request: PLATFileQueueRequest
        try {
          request = JSON.parse(await readFile(sourcePath, 'utf8')) as PLATFileQueueRequest
        } catch (error: any) {
          const invalidResponse: PLATFileQueueResponse = {
            id: basename(name, '.json'),
            ok: false,
            error: { status: 400, message: error?.message ?? 'Invalid file queue request' },
          }
          await writeFile(
            join(config.outbox, `${basename(name, '.json')}.response.json`),
            JSON.stringify(invalidResponse, null, 2),
          )
          await rm(sourcePath, { force: true })
          continue
        }

        const operation = runtime.resolveOperation({
          operationId: request.operationId,
          method: request.method,
          path: request.path,
        })

        if (!operation) {
          const notFoundResponse: PLATFileQueueResponse = {
            id: request.id,
            ok: false,
            error: { status: 404, message: `Operation not found for ${request.method} ${request.path}` },
          }
          await writeFile(
            join(config.outbox, `${request.id}.response.json`),
            JSON.stringify(notFoundResponse, null, 2),
          )
          await archiveRequest(sourcePath)
          continue
        }

        const eventsPath = join(config.outbox, `${request.id}.events.jsonl`)
        const normalizedInput = runtime.normalizeInput(request.input ?? {})

        const rt = runtime
        if (!rt) break

        const ctx: RouteContext = {
          method: request.method,
          url: request.path,
          headers: request.headers ?? {},
          opts: operation.routeMeta?.opts,
        }

        const abortController = new AbortController()
        rt.createCallContext({
          ctx,
          sessionId: request.id,
          mode: 'deferred',
          signal: abortController.signal,
          emit: async (event: 'progress' | 'log' | 'chunk' | 'message', data?: unknown) => {
            await writeFile(
              eventsPath,
              `${JSON.stringify({ id: request.id, event, data: rt.serializeValue(data) })}\n`,
              { flag: 'a' },
            )
          },
        })

        try {
          const envelope = rt.createEnvelope({
            protocol: 'file',
            operation,
            input: normalizedInput,
            headers: request.headers ?? {},
            ctx,
            requestId: request.id,
            req: { headers: request.headers ?? {} },
            allowHelp: false,
            helpRequested: false,
          })

          const execution = await rt.dispatch(operation, envelope)
          const response: PLATFileQueueResponse = {
            id: request.id,
            ok: true,
            result: rt.serializeValue(execution.result),
            statusCode: execution.statusCode,
          }
          await writeFile(
            join(config.outbox, `${request.id}.response.json`),
            JSON.stringify(response, null, 2),
          )
        } catch (error: any) {
          const status = error instanceof HttpError ? error.statusCode : 500
          const response: PLATFileQueueResponse = {
            id: request.id,
            ok: false,
            error: {
              status,
              message: error?.message ?? 'Internal server error',
              data: error instanceof HttpError ? error.data : undefined,
            },
          }
          await writeFile(
            join(config.outbox, `${request.id}.response.json`),
            JSON.stringify(response, null, 2),
          )
        }

        await archiveRequest(sourcePath)
      }
    } finally {
      busy = false
    }
  }

  return {
    name: 'file',

    setup(rt: PLATServerTransportRuntime) {
      runtime = rt
    },

    start() {
      if (!config || timer) return
      const pollIntervalMs = config.pollIntervalMs ?? 250
      void processOnce()
      timer = setInterval(() => { void processOnce() }, pollIntervalMs)
    },

    teardown() {
      if (timer) {
        clearInterval(timer)
        timer = undefined
      }
      runtime = undefined
    },
  }
}

File: python/plat/file_queue_protocol_plugin.py
from __future__ import annotations

import json
import os
import threading
import logging
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from .file_queue import FileQueueRequest, FileQueueSuccessResponse, FileQueueErrorResponse
from .protocol_plugin import ServerTransportRuntime
from .server_types import RouteContext

logger = logging.getLogger("plat")


@dataclass
class FileQueueProtocolPluginOptions:
    """Configuration for the file queue protocol plugin."""
    inbox: str
    outbox: str
    poll_interval_ms: int = 250
    archive: str | bool = True
    http_error_class: type | None = None


def create_file_queue_protocol_plugin(options: FileQueueProtocolPluginOptions) -> Any:
    """Create a self-contained file queue protocol plugin for the Python server."""

    HttpError = options.http_error_class

    runtime_ref: list[ServerTransportRuntime | None] = [None]
    stop_event = threading.Event()
    thread: list[threading.Thread | None] = [None]

    def archive_request(source_path: Path) -> None:
        if options.archive is False:
            source_path.unlink(missing_ok=True)
            return
        if isinstance(options.archive, str):
            source_path.rename(Path(options.archive) / source_path.name)
            return
        source_path.unlink(missing_ok=True)

    def process_once() -> None:
        import asyncio

        rt = runtime_ref[0]
        if rt is None:
            return

        os.makedirs(options.inbox, exist_ok=True)
        os.makedirs(options.outbox, exist_ok=True)
        if isinstance(options.archive, str):
            os.makedirs(options.archive, exist_ok=True)

        for name in sorted(entry for entry in os.listdir(options.inbox) if entry.endswith(".json")):
            source_path = Path(options.inbox) / name
            request_id = source_path.stem

            try:
                payload = json.loads(source_path.read_text(encoding="utf-8"))
                request = FileQueueRequest(**payload)
            except Exception as exc:
                response = FileQueueErrorResponse(
                    id=request_id,
                    ok=False,
                    error={"status": 400, "message": str(exc) or "Invalid file queue request"},
                )
                (Path(options.outbox) / f"{request_id}.response.json").write_text(
                    json.dumps(response.__dict__, indent=2), encoding="utf-8",
                )
                source_path.unlink(missing_ok=True)
                continue

            operation = rt.resolve_operation(
                operation_id=getattr(request, "operationId", None),
                method=request.method,
                path=request.path,
            )
            if operation is None:
                response = FileQueueErrorResponse(
                    id=request.id,
                    ok=False,
                    error={"status": 404, "message": f"Operation not found for {request.method} {request.path}"},
                )
                (Path(options.outbox) / f"{request.id}.response.json").write_text(
                    json.dumps(response.__dict__, indent=2), encoding="utf-8",
                )
                archive_request(source_path)
                continue

            events_path = Path(options.outbox) / f"{request.id}.events.jsonl"
            input_data = rt.normalize_input(dict(request.input or {}))

            ctx = RouteContext(
                method=request.method,
                url=request.path,
                headers=dict(request.headers or {}),
                opts=operation.route_meta.opts if operation.route_meta else None,
            )

            def make_emit(req_id: str, ev_path: Path):
                def emit(event: str, data: Any = None) -> None:
                    line = json.dumps({
                        "id": req_id, "event": event,
                        "data": rt.serialize_value(data),
                    }) + "\n"
                    with open(ev_path, "a", encoding="utf-8") as f:
                        f.write(line)
                return emit

            emit_fn = make_emit(request.id, events_path)
            rt.create_call_context(
                ctx=ctx,
                session_id=request.id,
                mode="deferred",
                emit=emit_fn,
                signal=None,
            )

            try:
                execution = asyncio.run(
                    rt.dispatch(
                        operation,
                        rt.create_envelope(
                            protocol="file",
                            operation=operation,
                            input=input_data,
                            ctx=ctx,
                            headers=dict(request.headers or {}),
                            request_id=request.id,
                            request=type("FileQueueRequest", (), {"headers": dict(request.headers or {})})(),
                            allow_help=False,
                            help_requested=False,
                        ),
                    )
                )
                response = FileQueueSuccessResponse(
                    id=request.id,
                    ok=True,
                    result=rt.serialize_value(execution["result"]),
                    statusCode=execution["status_code"],
                )
                (Path(options.outbox) / f"{request.id}.response.json").write_text(
                    json.dumps(response.__dict__, indent=2), encoding="utf-8",
                )
            except Exception as exc:
                status = 500
                if HttpError and isinstance(exc, HttpError):
                    status = exc.status_code
                else:
                    status = getattr(exc, "status_code", None) or getattr(exc, "status", None) or 500
                response = FileQueueErrorResponse(
                    id=request.id,
                    ok=False,
                    error={
                        "status": status,
                        "message": str(exc) or "Internal server error",
                        **({"data": getattr(exc, "data")} if getattr(exc, "data", None) is not None else {}),
                    },
                )
                (Path(options.outbox) / f"{request.id}.response.json").write_text(
                    json.dumps(response.__dict__, indent=2), encoding="utf-8",
                )

            archive_request(source_path)

    class FileQueueProtocolPlugin:
        name = "file"

        def setup(self, runtime: ServerTransportRuntime) -> None:
            runtime_ref[0] = runtime

        def start(self, runtime: ServerTransportRuntime) -> None:
            if thread[0] is not None:
                return
            poll_seconds = max(0.05, options.poll_interval_ms / 1000)

            def worker() -> None:
                while not stop_event.is_set():
                    try:
                        process_once()
                    except Exception as exc:
                        logger.error("File queue processing failed: %s", exc)
                    stop_event.wait(poll_seconds)

            t = threading.Thread(target=worker, name="plat-file-queue", daemon=True)
            thread[0] = t
            t.start()

        def teardown(self, runtime: ServerTransportRuntime) -> None:
            stop_event.set()
            if thread[0] is not None:
                thread[0].join(timeout=5)
                thread[0] = None
            runtime_ref[0] = None

        def process_once(self) -> None:
            process_once()

    return FileQueueProtocolPlugin()
